lib/filterx/func-unset-empties.c

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) 2024 Attila Szakacs
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filterx/func-unset-empties.h"
+#include "filterx/object-string.h"
+#include "filterx/object-primitive.h"
+#include "filterx/object-null.h"
+#include "filterx/object-dict-interface.h"
+#include "filterx/object-list-interface.h"
+#include "filterx/filterx-eval.h"
+#include "filterx/filterx-globals.h"
+
+#define FILTERX_FUNC_UNSET_EMPTIES_USAGE "Usage: unset_empties(object, recursive=true, string=true, null=true, number=true, dict=true, list=true)"
+
+typedef struct FilterXFunctionUnsetEmpties_
+{
+  FilterXFunction super;
+  FilterXExpr *object_expr;
+  gboolean recursive;
+  gboolean drop_strings;
+  gboolean drop_nulls;
+  gboolean drop_numbers;
+  gboolean drop_dicts;
+  gboolean drop_lists;
```

Comment:
What's our exact use-case with implementing `unset_empties()` with type-based parameters (filters)?

I'm asking because these bool parameters don't really match with clean-code principles (from filterx perspective), and adding a bool field in C for each scalar types also looks a bit weird. `_should_unset()` defaults to false, so new scalar types will also be ignored (not unset) if we forget about this implementation here.

I think a better approach would be to receive a filter expression for `unset_empties()`, or even better, let `unset_empties()` unset all empties without condition, and implement a separate function that can do transformations on dict and list values, allowing filtering, setting, unsetting (like map, reduce(), etc.).